### PR TITLE
Add back in the option of using an external core

### LINKED ISSUE #121: Add back in the option of using an external core
This option was removed in #119.  Requested at https://github.com/editorconfig/editorconfig-vim/pull/119#issuecomment-498837689 .

@xuhdev Did a user request this feature?  Could you tell me a bit more about the use case?  I am confused.  Thanks!

### PR DESCRIPTION
**Pull request by cxw42**
Closes #121

### CODE REVIEW

**Comment by xuhdev** [comment]
Thanks. Is this PR a simple revertion of a previous commit? Or did you adjust a bit?

**Comment by cxw42** [comment]
It is not a straight reversion.  Previously, this plugin supported a bundled Python core or an external core.  I added back in the code for the external core.  I did not add the code for the Python core since that has been replaced by the Vimscript core.

Because there are fewer core options, I was able to simplify some of the logic.  I also added fallback from `external_command` to `vim_core` so that the plugin will still work even if the `external_command` cannot be initialized.

For reference, [here](https://github.com/editorconfig/editorconfig-vim/compare/68f8136d2b018bfa9b23403e87d3d65bc942cbc3...b03ab7a759a4f40a8f1b43c6cf065e5715e29f56) is a comparison between the state just before the switch to the vim core and this PR.

File: plugin/editorconfig.vim [hunk 1]
```diff
@@ -107,50 +117,56 @@ function! s:InitializeVimCore()
...
+function! s:InitializeExternalCommand()
+" Initialize external_command mode
+
+    if empty(s:editorconfig_exec_path)
+        echo 'Please specify a g:EditorConfig_exec_path'
+        return 1
+    endif
+
+    if g:EditorConfig_verbose
+        echo 'Checking for external command ' . s:editorconfig_exec_path
```
Comment (xuhdev) [inline plugin/editorconfig.vim]: nit: `. ' ...'`

File: plugin/editorconfig.vim [hunk 2]
```diff
@@ -107,50 +117,56 @@ function! s:InitializeVimCore()
...
-            if s:InitializeVimCore()
-                echo 'EditorConfig: Failed to initialize vim_core mode'
-                return 1
-            endif
-        else
-            echo "EditorConfig: I don't know how to use mode " . s:editorconfig_core_mode
-            return 1
+    if s:editorconfig_core_mode ==? 'external_command'
+        if s:InitializeExternalCommand()
+            echo 'EditorConfig: Failed to initialize external_command mode.  ' .
```
Comment (xuhdev) [inline plugin/editorconfig.vim]: Because our default is vim_core mode, likely this is something unexpected to user. Can we do a warning color?

```vim
  echohl WarningMsg
  echo "Warning message"
  echohl None
```

File: plugin/editorconfig.vim [hunk 3]
```diff
@@ -107,50 +117,56 @@ function! s:InitializeVimCore()
...
-
-    " Determine the editorconfig_core_mode we should use {{{2
-    while 1
-        " If user has specified a mode, just break
-        if exists('s:editorconfig_core_mode') && !empty(s:editorconfig_core_mode)
-            break
+    if s:editorconfig_core_mode ==? 'vim_core'
+        if s:InitializeVimCore()
+            echo 'EditorConfig: Failed to initialize vim_core mode.  ' .
+                \ 'The plugin will not function.'
```
Comment (xuhdev) [inline plugin/editorconfig.vim]: warning message

File: plugin/editorconfig.vim [hunk 4]
```diff
@@ -107,50 +117,56 @@ function! s:InitializeVimCore()
...
+        " Nothing to do here, but this elseif is required to avoid
+        " external_command falling into the else clause.
 
-    " No EditorConfig Core is available
-    if empty(s:editorconfig_core_mode)
-        echo "EditorConfig: ".
-                    \ "No EditorConfig Core is available.  " .
-                    \ "The plugin won't work."
+    else    " neither external_command nor vim_core
+        echo "EditorConfig: I don't know how to use mode " . s:editorconfig_core_mode
```
Comment (xuhdev) [inline plugin/editorconfig.vim]: Error message

File: doc/editorconfig.txt [hunk 5]
```diff
@@ -86,6 +91,21 @@ Example: Avoid loading EditorConfig for any remote files over ssh
...
+>
+ let g:EditorConfig_exec_path = 'Path to your EditorConfig Core executable'
+<
+The default value is empty.
+
+If "g:EditorConfig_exec_path" is not set, the plugin will use the "vim_core"
+mode regardless of the setting of |g:EditorConfig_core_mode|.
+
+Changes to "g:EditorConfig_exec_path" will not take effect until Vim
+is restarted.
```
Comment (xuhdev) [inline doc/editorconfig.txt]: There used to be a path search program here to search for default location:

https://github.com/editorconfig/editorconfig-vim/compare/68f8136d2b018bfa9b23403e87d3d65bc942cbc3...b03ab7a759a4f40a8f1b43c6cf065e5715e29f56#diff-b58403939457354bd5850ccec1b0c8e3L111

Do you think it's worth being added back?

Comment (cxw42) [reply doc/editorconfig.txt]: A good question.  I do not think it is worth the complication for what is now a backup option.  I suggest that a user who knows enough to acquire a core and set this option knows enough to provide the full path :) .  Removing path search also increases portability.

Comment (xuhdev) [reply doc/editorconfig.txt]: Sounds good. Maybe default to `which editorconfig`?

Comment (cxw42) [reply doc/editorconfig.txt]: I can, but I must honestly tell you I have no desire to do so :) .  

- Any form of automatic search adds code and complication.
- Platform-specific code (which path search is) is harder to test and more likely to fail in edge cases.
- I am not an expert on `shellescape`.
- `which` doesn't exist on all platforms.  Vim supports MS-DOS, which has neither `which` nor `where`.

This is an option almost no-one will use.  Anyone who knows enough to use this option knows enough to add two lines to `.vimrc` instead of one.

I will update the documentation for `g:EditorConfig_core_mode` to note that you have to set `g:EditorConfig_exec_path` if you are using `external_command` mode.  What do you think about that?

Comment (xuhdev) [reply doc/editorconfig.txt]: Sounds reasonable to me. Thanks!

**Review by M3T4F34R — COMMENTED** [review]
editorconfig.editorconfig
